File: 0229-majority-element-ii/0229-majority-element-ii.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        int count1 = 0, count2 = 0;
        int ele1 = 0, ele2 = 0;

        for(int num : nums) {
            if(count1 == 0 && num != ele2) {
                count1++;                   //count1 = 2, ele1 = 3, count2 = 1, ele2 = 2
                ele1 = num;
            }
            else if(count2 == 0 && num != ele1) {
                count2++;
                ele2 = num;
            }
            else if(num == ele1) {
                count1++;
            }
            else if(num == ele2) {
                count2++;
            }
            else {
                count1--;
                count2--;
            }
        }

        count1 = 0;
        count2 = 0;
        for(int num : nums) {
            if(num == ele1) count1++;
            if(num == ele2) count2++;
        }

        int mini = nums.length / 3;
        List<Integer> res = new ArrayList<>();
        if(count1 > mini) res.add(ele1);
        if(ele2 != ele1 && count2 > mini) res.add(ele2);

        return res;
 
    }
}

File: 0229-majority-element-ii/0229-majority-element-ii.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        vector<int> ans;
        int cnt1 = 0, cnt2 = 0;
        int n = nums.size();
        int el1 = INT_MIN, el2 = INT_MIN;
        
        for(int i=0; i<n; i++) {
            if(cnt1 == 0 && el2 != nums[i]) {
                cnt1 = 1;
                el1 = nums[i];
            }
            
            else if(cnt2 == 0 && el1 != nums[i]) {
                cnt2 = 1;
                el2 = nums[i];
            }
            
            else if(el1 == nums[i]) cnt1++;
            else if(el2 == nums[i]) cnt2++;
            
            else  {
                cnt1--;
                cnt2--;
                
            }
            
        }
        
        cnt1 = 0, cnt2 = 0;
        
        for(int i=0; i<n; i++) {
            if(el1 == nums[i]) cnt1++;
            if(el2 == nums[i]) cnt2++;
        }
        
        if(cnt1 > n/3) {
            ans.push_back(el1);
        }
        
        if(cnt2 > n/3) {
            ans.push_back(el2);
        }
        
        return ans;
        
    }
};